File: app/src/main/java/com/gnoemes/shikimoriapp/data/repository/anime/AnimeRepositoryImpl.java
package com.gnoemes.shikimoriapp.data.repository.anime;

import com.gnoemes.shikimoriapp.data.local.db.HistoryDbSource;
import com.gnoemes.shikimoriapp.data.local.db.RateSyncDbSource;
import com.gnoemes.shikimoriapp.data.network.AnimesApi;
import com.gnoemes.shikimoriapp.data.repository.anime.converter.AnimeDetailsResponseConverter;
import com.gnoemes.shikimoriapp.data.repository.anime.converter.AnimeListResponseConverter;
import com.gnoemes.shikimoriapp.data.repository.anime.converter.FranchiseResponseConverter;
import com.gnoemes.shikimoriapp.data.repository.anime.converter.LinkResponseConverter;
import com.gnoemes.shikimoriapp.data.repository.anime.converter.ScreenshotResponseConverter;
import com.gnoemes.shikimoriapp.entity.anime.domain.Anime;
import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeDetails;
import com.gnoemes.shikimoriapp.entity.anime.domain.FranchiseNode;
import com.gnoemes.shikimoriapp.entity.anime.domain.Link;
import com.gnoemes.shikimoriapp.entity.anime.series.data.db.HistoryDao;
import com.gnoemes.shikimoriapp.entity.screenshots.domain.Screenshot;

import org.joda.time.DateTimeComparator;

import java.util.LinkedHashSet;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Single;

public class AnimeRepositoryImpl implements AnimeRepository {

    private AnimesApi animesApi;
    private RateSyncDbSource syncDbSource;
    private HistoryDbSource historyDbSource;
    private AnimeDetailsResponseConverter responseConverter;
    private AnimeListResponseConverter listResponseConverter;
    private LinkResponseConverter linkReponseConverter;
    private FranchiseResponseConverter franchiseResponseConverter;
    private ScreenshotResponseConverter screenshotResponseConverter;

    @Inject
    public AnimeRepositoryImpl(AnimesApi animesApi,
                               RateSyncDbSource syncDbSource,
                               HistoryDbSource historyDbSource,
                               AnimeDetailsResponseConverter responseConverter,
                               LinkResponseConverter linkReponseConverter,
                               AnimeListResponseConverter listResponseConverter,
                               FranchiseResponseConverter franchiseResponseConverter,
                               ScreenshotResponseConverter screenshotResponseConverter) {
        this.animesApi = animesApi;
        this.syncDbSource = syncDbSource;
        this.historyDbSource = historyDbSource;
        this.responseConverter = responseConverter;
        this.linkReponseConverter = linkReponseConverter;
        this.listResponseConverter = listResponseConverter;
        this.franchiseResponseConverter = franchiseResponseConverter;
        this.screenshotResponseConverter = screenshotResponseConverter;
    }

    @Override
    public Single<AnimeDetails> getAnimeDetails(long animeId) {
        return animesApi.getAnimeDetails(animeId)
                .flatMap(animeDetailsResponse -> animesApi.getRoles(animeId)
                        .map(rolesResponses -> responseConverter.convertDetailsWithCharacters(animeDetailsResponse, rolesResponses)))
                .flatMap(animeDetails -> Single.just(animeDetails)
                        .filter(details -> details.getAnimeRate() != null)
                        .flatMapCompletable(anime -> syncDbSource.saveRateEpisodes(
                                anime.getAnimeRate().getId(),
                                anime.getId(),
                                Integer.parseInt(anime.getAnimeRate().getEpisodes())))
                        .toSingleDefault(animeDetails));
    }

    @Override
    public Single<List<Link>> getAnimeLinks(long animeId) {
        return animesApi.getAnimeLinks(animeId)
                .map(linkReponseConverter);
    }

    @Override
    public Single<List<Anime>> getSimilarAnimes(long animeId) {
        return animesApi.getSimilarAnimes(animeId)
                .map(listResponseConverter);
    }

    @Override
    public Single<List<FranchiseNode>> getFranchiseNodes(long animeId) {
        return animesApi.getFranchise(animeId)
                .map(franchiseResponseConverter)
                .flatMap(nodes -> Observable.fromIterable(nodes)
                        .toSortedList((o1, o2) -> DateTimeComparator.getInstance().compare(o1.getDateTime(), o2.getDateTime())));
    }

    @Override
    public Single<List<Screenshot>> getScreenshots(long animeId) {
        return animesApi.getScreenshots(animeId)
                .map(screenshotResponseConverter);
    }

    @Override
    public Single<LinkedHashSet<Long>> getLocalWatchedAnimeIds() {
        return historyDbSource.getWatchedAnimes()
                .flatMap(historyDaos -> Observable.fromIterable(historyDaos)
                        .map(HistoryDao::getAnimeId)
                        .toList())
                .map(LinkedHashSet::new);
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/data/repository/anime/AnimeRepositoryImpl.kt
package com.gnoemes.shikimoriapp.data.repository.anime

import com.gnoemes.shikimoriapp.data.local.db.HistoryDbSource
import com.gnoemes.shikimoriapp.data.local.db.RateSyncDbSource
import com.gnoemes.shikimoriapp.data.network.AnimesApi
import com.gnoemes.shikimoriapp.data.repository.anime.converter.AnimeDetailsResponseConverter
import com.gnoemes.shikimoriapp.data.repository.anime.converter.AnimeFranchiseResponseConverter
import com.gnoemes.shikimoriapp.data.repository.anime.converter.AnimeLinkResponseConverter
import com.gnoemes.shikimoriapp.data.repository.anime.converter.AnimeListResponseConverter
import com.gnoemes.shikimoriapp.entity.anime.domain.Anime
import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeDetails
import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeFranchiseNode
import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeLink
import io.reactivex.Observable
import io.reactivex.Single
import org.joda.time.DateTimeComparator
import java.util.*
import javax.inject.Inject

class AnimeRepositoryImpl @Inject constructor(
        val api: AnimesApi,
        val syncDbSource: RateSyncDbSource,
        val historyDbSource: HistoryDbSource,
        val responseConverter: AnimeDetailsResponseConverter,
        val listResponseConverter: AnimeListResponseConverter,
        val linkResponseConverter: AnimeLinkResponseConverter,
        val franchiseResponseConverter: AnimeFranchiseResponseConverter
) : AnimeRepository {


    override fun getAnimeDetails(animeId: Long): Single<AnimeDetails> =
            api.getAnimeDetails(animeId)
                    .flatMap { details ->
                        api.getRoles(animeId)
                                .map { responseConverter.convertDetailsWithCharacters(details, it) }
                    }
                    .flatMap { details ->
                        Single.just(details)
                                .filter { it.userRate != null }
                                .flatMapCompletable { anime ->
                                    syncDbSource.saveRateEpisodes(
                                            anime.userRate!!.id,
                                            anime.id,
                                            anime.userRate.episodes!!.toInt())
                                }
                                .toSingleDefault(details)
                    }


    override fun getAnimeLinks(animeId: Long): Single<MutableList<AnimeLink>> =
            api.getAnimeLinks(animeId)
                    .map(linkResponseConverter)

    override fun getSimilarAnimes(animeId: Long): Single<MutableList<Anime>> =
            api.getSimilarAnimes(animeId)
                    .map(listResponseConverter)

    override fun getFranchiseNodes(animeId: Long): Single<MutableList<AnimeFranchiseNode>> =
            api.getFranchise(animeId)
                    .map(franchiseResponseConverter)
                    .flatMap { nodes ->
                        Observable.fromIterable<AnimeFranchiseNode>(nodes)
                                .toSortedList { o1, o2 -> DateTimeComparator.getInstance().compare(o1.dateTime, o2.dateTime) }
                    }

    override fun getLocalWatchedAnimeIds(): Single<LinkedHashSet<Long>> =
            historyDbSource.watchedAnimes
                    .flatMap { historyDaos ->
                        Observable.fromIterable(historyDaos)
                                .map { it.animeId }
                                .toList()
                    }
                    .map { LinkedHashSet(it) }
}